File: 프로그래머스/lv2/42842. 카펫/카펫.py
import math
#약수는 대칭적인 관계니까, 굳이 끝까지 탐색할 필요 없이 중간까지만 봐도됨
def solution(brown, yellow):
    total = brown + yellow
    total_middle = int(math.sqrt(total)) + 1
    for i in range(1, total_middle):
        if total % i == 0:
            width = total // i
            height = i
            if (width - 2) * (height - 2) == yellow:
                return [width, height]

File: 프로그래머스/2/42842. 카펫/카펫.java
import java.util.*;

class Solution {
    public int[] solution(int brown, int yellow) {
        int[] answer = new int[2];
        int carpetSize = brown + yellow;
        
        Deque<Integer> deque = findWidthAndHeight(carpetSize);
        while (true) {
            int width = deque.pollFirst();
            int height = deque.pollFirst();
            
            if((width - 2) * (height - 2) == yellow) {
                answer[0] = width;
                answer[1] = height;
                break;
            }
        }
        
        int temp = 0;
        if (answer[1] > answer[0]) {
            temp = answer[0];
            answer[0] = answer[1];
            answer[1] = temp;
        }
        
        return answer;
    }
    
    private Deque<Integer> findWidthAndHeight(int size) {
        Deque<Integer> deque = new ArrayDeque<>();
        
        int sqrtSize = (int)Math.sqrt(size);
        int height = 3;
        while(true) {
            if (height > sqrtSize) {
                break;
            }

            if (size % height == 0) {
                int width = size / height;
                deque.addLast(width);
                deque.addLast(height);
            }
            height ++;
        }
        
        return deque;
    }
}